### ISSUE TITLE
Preview 7 Improper LinkTarget returned for a Directory Junction on Windows and Inconsistent FinalTarget results

### ISSUE THREAD

**Issue by fitdev** [dotnet/runtime#57575]
## Main Issue

When a directory is a junction (as opposed to being a regular directory or a symlink) on Windows, as of Preview 7, `DirectoryInfo.LinkTarget` returns a non-null value, which is wrong. Considering a structure like this:

```
C:\Test\RealFolder
C:\Test\JunctionToRealFolder -> C:\Test\RealFolder
C:\Test\SymlinkToRealFolder -> C:\Test\RealFolder
```

`DirectoryInfo.LinkTarget` for `C:\Test\JunctionToRealFolder` returns `\Test\RealFolder` (instead of `C:\Test\RealFolder`)
`DirectoryInfo.LinkTarget` for `C:\Test\SymlinkToRealFolder ` returns `C:\Test\RealFolder` (which _is correct_)

Strangely, `DirectoryInfo.ResolveLinkTarget(returnFinalTarget: true)` returns the correct target in all cases (Junction and Symlink).

I suspect the issue comes that you have an incorrect check for a reparse tag kind in your code:
https://github.com/dotnet/runtime/blob/eeee548ba69083cce4c2a8cfb9f7222355b2d1a0/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs#L506

Should instead be:

```cs
if (rdb.ReparseTag != Interop.Kernel32.IOReparseOptions.IO_REPARSE_TAG_SYMLINK)
```

Since you say you support _symlinks only_ ATM.

## Smaller issue

On Windows, there is an inconsistency of final targets returned depending on which APIs are used, and whether or not virtual drives created with [subst command](https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/subst) are used in the path (both the link and the target residing on such virtual drives):

The `DirectoryInfo.ResolveLinkTarget(returnFinalTarget: true)` API (and other APIs you added that resolve _final_ target) resolve the final target to _non-virtual_ drives, while the `DirectoryInfo.LinkTarget` (and other similar APIs that resolve the immediate target), when _called repeatedly while LinkTarget != null_ stop at the virtual drive level.

So, given a structure like this:

```
C:\RootDirectoryForVirtualDriveX
X:\SomeRealFolder
X:\SymlinkToRealFolderOnX -> X:\SomeRealFolder
```

Calling `DirectoryInfo.ResolveLinkTarget(returnFinalTarget: true)` on `X:\SymlinkToRealFolderOnX` will result in: `C:\RootDirectoryForVirtualDriveX\SomeRealFolder`

While calling `DirectoryInfo.LinkTarget` (repeatedly while the result is non-null) on `X:\SymlinkToRealFolderOnX` will result in: `X:\SomeRealFolder`

Even though, logically, and for consistency results should be the same.

**Comment by adamsitnik** [comment]
@carlossanlop @Jozkee PTAL

**Comment by carlossanlop** [comment]
Thanks for your report, @fitdev

The "main issue" should get moved to 7.0. As you correctly stated, we only support symlinks at the moment.

I'm moving the "smaller issue" into a separate issue, and we can then determine if it should be fixed for 6.0 or 7.0.

**Comment by fitdev** [comment]
@carlossanlop Thanks for looking into this. However this issue was not about _supporting_ Junctions. It was about a bug in Preview 7 and your current code that basically returns a target (i.e. non-null string) even if the directory _is_ a Junction. Instead it should return null then, and _certainly NOT_ the totally erroneous path as it does currently.

**Comment by carlossanlop** [comment]
Good point, @fitdev. I'll mark it for 6.0 and return null for junctions.

**Comment by carlossanlop** [comment]
@fitdev I had a conversation with @Jozkee and there's no need to return `null` in `LinkTarget` if the link is a junction. The fix ensures both public APIs `ResolveLinkTarget(bool)` and `LinkTarget` can resolve both symlinks and junctions, because internally, they both call the same method: `GetImmediateLinkTarget`.
Even if currently we don't offer a public API to determine what kind of link the file represents, the user wants to be able to succesfully resolve the target. This scenario is much more important.

**Comment by fitdev** [comment]
If it would work correctly, I suppose. But currently it does not work correctly, see my original issue comment: 

> DirectoryInfo.LinkTarget for C:\Test\JunctionToRealFolder returns \Test\RealFolder (instead of C:\Test\RealFolder)

Better to just return null until Junctions are properly supported instead of an improper path like this.

**Comment by carlossanlop** [comment]
Noted. With the bug fix, the `LinkTarget` API should now return the correct path for your example: `C:\Test\RealFolder`.

**Comment by fitdev** [comment]
I noticed you merged your changes into main yesterday. However, I see that you still use the old check to determine reparse tag kind: https://github.com/dotnet/runtime/blob/cc0c1213003ca9fbbae5c300c6ac47738f4a3fd7/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs#L559-L560

Shouldn't these be:

```cs
 ((data.dwReserved0 != Interop.Kernel32.IOReparseOptions.IO_REPARSE_TAG_SYMLINK) && 
  (data.dwReserved0 != Interop.Kernel32.IOReparseOptions.IO_REPARSE_TAG_MOUNT_POINT))) 
```

I mean otherwise, supposing reparse tag was say `0x80000021` (`IO_REPARSE_TAG_ONEDRIVE`), then `0x80000021 & 0xA000000C` would be `0x80000000`, which is _not_ `0`! Therefore routine will _not_ return `null` (as it should have, since reparse tag is neither `IO_REPARSE_TAG_SYMLINK` or `IO_REPARSE_TAG_MOUNT_POINT`).

**Comment by Jozkee** [comment]
> I mean otherwise, supposing reparse tag was say 0x80000021 (IO_REPARSE_TAG_ONEDRIVE), then 0x80000021 & 0xA000000C would be 0x80000000, which is not 0! Therefore routine will not return null (as it should have, since reparse tag is neither IO_REPARSE_TAG_SYMLINK or IO_REPARSE_TAG_MOUNT_POINT).

The fix for that will be in .NET 6-RC2 release. 
Thanks again, @fitdev.
